File: companion/lib/transcriptClaude.js
const { isSyntheticUserMessage, textOf } = require('./syntheticUser');

function toolArg(name, input) {
  const i = input || {};
  if (name === 'Read' || name === 'Edit' || name === 'Write') {
    const segs = String(i.file_path || '').split('/').filter(Boolean);
    return segs.slice(-2).join('/');
  }
  if (name === 'Bash') return String(i.command || '').split('\n')[0].slice(0, 40);
  if (name === 'Grep' || name === 'Glob') return String(i.pattern || '');
  return '';
}

function parseClaudeTranscript(text) {
  const lines = String(text || '').split('\n');
  const r = {
    totalInput: 0, totalOutput: 0, totalCacheRead: 0, totalCacheCreate: 0,
    lastContextTokens: 0, maxContextTokens: 0, turnCount: 0,
    model: '', version: '', gitBranch: '', currentTask: '', firstPrompt: '',
    lastUserTsMs: 0, pendingTool: false,
  };
  for (const line of lines) {
    const s = line.trim();
    if (!s) continue;
    let e;
    try { e = JSON.parse(s); } catch (_) { continue; }      // 坏行跳过

    if (e.type === 'assistant' && e.message) {
      const u = e.message.usage || {};
      const inp = Number(u.input_tokens) || 0;
      const out = Number(u.output_tokens) || 0;
      const cr = Number(u.cache_read_input_tokens) || 0;
      const cc = Number(u.cache_creation_input_tokens) || 0;
      r.totalInput += inp; r.totalOutput += out; r.totalCacheRead += cr; r.totalCacheCreate += cc;
      const ctx = (cr === 0 && cc > 0) ? (inp + cc) : (inp + cr);
      r.lastContextTokens = ctx;
      if (ctx > r.maxContextTokens) r.maxContextTokens = ctx;
      if (e.message.model) r.model = e.message.model;
      r.turnCount += 1;
      // currentTask = 本轮最后一个 tool_use(新轮覆盖)
      let task = '';
      let hasTool = false;
      for (const b of (e.message.content || [])) {
        if (b && b.type === 'tool_use') { hasTool = true; task = (b.name + ' ' + toolArg(b.name, b.input)).trim(); }
      }
      r.currentTask = task;
      r.pendingTool = hasTool;           // 末尾若是该 assistant 则仍 pending(后续 user 行会覆盖)
      r.lastUserTsMs = 0;                // assistant 之后清除 user 等待标记
    } else if (e.type === 'user') {
      if (e.version) r.version = e.version;
      if (e.gitBranch) r.gitBranch = e.gitBranch;
      const synthetic = isSyntheticUserMessage(e);
      if (!synthetic && !r.firstPrompt) r.firstPrompt = textOf(e.message && e.message.content).trim();
      r.pendingTool = false;             // user 行(含 tool_result)闭合上一个 assistant 的工具
      r.lastUserTsMs = synthetic ? 0 : (Date.parse(e.timestamp) || 0);
    }
  }
  return r;
}

module.exports = { parseClaudeTranscript, toolArg };

File: companion/codey/transcript_claude.py
"""解析 Claude Code transcript JSONL -> 聚合对象。纯函数(入参为字符串)。"""
import json

from .synthetic import is_synthetic_user_message, text_of
from .util import parse_ts_ms


def tool_arg(name, inp):
    i = inp or {}
    if name in ("Read", "Edit", "Write"):
        segs = [p for p in str(i.get("file_path") or "").split("/") if p]
        return "/".join(segs[-2:])
    if name == "Bash":
        return str(i.get("command") or "").split("\n")[0][:40]
    if name in ("Grep", "Glob"):
        return str(i.get("pattern") or "")
    return ""


def _int(v):
    try:
        return int(v or 0)
    except (TypeError, ValueError):
        return 0


def parse_claude_transcript(text):
    r = {
        "total_input": 0, "total_output": 0, "total_cache_read": 0, "total_cache_create": 0,
        "last_context_tokens": 0, "max_context_tokens": 0, "turn_count": 0,
        "model": "", "version": "", "git_branch": "", "current_task": "", "first_prompt": "",
        "last_user_ts_ms": 0, "pending_tool": False,
        "compactions": 0, "_prev_ctx": 0,
    }
    for line in str(text if text is not None else "").split("\n"):
        s = line.strip()
        if not s:
            continue
        try:
            e = json.loads(s)
        except (ValueError, TypeError):
            continue
        if not isinstance(e, dict):
            continue
        t = e.get("type")
        if t == "assistant" and e.get("message"):
            msg = e["message"]
            u = msg.get("usage") or {}
            inp, out = _int(u.get("input_tokens")), _int(u.get("output_tokens"))
            cr, cc = _int(u.get("cache_read_input_tokens")), _int(u.get("cache_creation_input_tokens"))
            r["total_input"] += inp
            r["total_output"] += out
            r["total_cache_read"] += cr
            r["total_cache_create"] += cc
            ctx = (inp + cc) if (cr == 0 and cc > 0) else (inp + cr)
            if r["_prev_ctx"] > 0 and ctx < r["_prev_ctx"] * 0.7:
                r["compactions"] += 1
            r["_prev_ctx"] = ctx
            r["last_context_tokens"] = ctx
            if ctx > r["max_context_tokens"]:
                r["max_context_tokens"] = ctx
            if msg.get("model"):
                r["model"] = msg["model"]
            r["turn_count"] += 1
            task, has_tool = "", False
            for b in (msg.get("content") or []):
                if isinstance(b, dict) and b.get("type") == "tool_use":
                    has_tool = True
                    task = (str(b.get("name") or "") + " " + tool_arg(b.get("name"), b.get("input"))).strip()
            r["current_task"] = task
            r["pending_tool"] = has_tool
            r["last_user_ts_ms"] = 0
        elif t == "user":
            if e.get("version"):
                r["version"] = e["version"]
            if e.get("gitBranch"):
                r["git_branch"] = e["gitBranch"]
            synthetic = is_synthetic_user_message(e)
            if not synthetic and not r["first_prompt"]:
                r["first_prompt"] = text_of((e.get("message") or {}).get("content")).strip()
            r["pending_tool"] = False
            r["last_user_ts_ms"] = 0 if synthetic else parse_ts_ms(e.get("timestamp"))
    r.pop("_prev_ctx", None)
    return r
